### DISCUSSION TITLE
Standard parallel algorithms on a custom queue

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1575] [Q&A] [answered]
Hello,

Thank you for the excellent SYCL implementation! I have a question regarding the use of custom queues with standard parallel algorithms. Currently, these algorithms execute on their own queue, but I need them to run on a queue that I have created on a specific device chosen in runtime by the user.

Is there a way to specify a custom queue for these standard parallel algorithms? Alternatively, is it be possible to access and use the implementation of these algorithms by other way, may be using the AdaptiveCpp sources?

Thank you!

**Comment by illuhad** [comment, marked as answer]
If you need more control, what you can do is to access the lower-level `hipsycl::algorithms` library directly. These library functions mostly map 1:1 to the standard parallel algorithms, but their signature is slightly different:
- They have `sycl::queue&` as first argument
- They return `sycl::event`
- Some of them have some additional arguments to manage scratch memory that the algorithms might need.


Note that at some point in the future the `hipsycl::algorithms` namespace will be renamed to `acpp::algorithms` or similar. So maybe it is a good idea for you to introduce an alias here so that you only need to change one place in your code ;)

Currently I would strongly recommend using these algorithms exclusively with in-order queues. They are untested with out-of-order queues, and the signature of many algorithms currently lacks dependency events. So it is currently difficult to specify asynchronous dependencies.

Note that when you use custom queues and submit to your own set of runtime-selected devices, you can no longer rely on the automatic C++ memory management interception from `--acpp-stdpar` to work. Replacing C++ memory allocations with SYCL memory management as `--acpp-stdpar` does requires deciding on one particular device (or at least backend) as soon as the application starts up and starts allocating memory.
So you will have to do all memory allocations using explicit SYCL functionality, like in normal SYCL programs.

**Comment by blinkfrog** [comment]
Amazing, thanks!
